models/intermediate/int_sap__0employee_date_changes.sql

```diff
@@ -0,0 +1,108 @@
+{% set employee_models = ['pa0000', 'pa0001', 'pa0007', 'pa0008', 'pa0031'] %}
+
+with
+{% for employee_model in employee_models %}
+    {{ employee_model }}_beg as (
+        select
+            pernr,
+            begda as date_change
+        from {{ var(employee_model) }}
+    ),
+
+    {{ employee_model }}_end as (
+        select
+            pernr,
+			case when endda = '99991231' then endda
+				else
+				--convert date strings to dates, add one day, and back to avoid null records as we would join on date ranges in the end model that don't exist
+				{% if target.type in ('redshift', 'snowflake', 'snowflake-sap') %}
+				to_char(dateadd(day, 1, to_date(endda, 'YYYYMMDD')), 'YYYYMMDD') end as date_change
+				{% elif target.type == 'postgres' %}
+				to_char(to_date(endda, 'YYYYMMDD') + interval '1 day', 'YYYYMMDD') end as date_change
+				{% elif target.type == 'bigquery' %}
+				format_date('%Y%m%d', date_add(parse_date('%Y%m%d', endda), interval 1 day)) end as date_change
+				{% elif target.type == 'databricks' %}
+				date_format(date_add(to_date(endda, 'yyyyMMdd'), 1), 'yyyyMMdd') end as date_change
+				{% endif %}
+		from {{ var(employee_model) }}
+	), 
+{% endfor %}
+
+unioned_dates as (
+
+	{% for employee_model in employee_models %}
+	
+		{% if target.type == 'bigquery' %}
+		select 
+			pernr,
+			date_change 
+		from {{ employee_model }}_beg
+		{{ dbt_utils.group_by(2) }}
+
+		union distinct
+
+		select 
+			pernr,
+			date_change 
+		from {{ employee_model }}_end
+		{{ dbt_utils.group_by(2) }}
+
+			{% if not loop.last %}
+    		union distinct
+			{% endif %} 
+
+		{% else %}
+		select 
+			pernr,
+			date_change 
+		from {{ employee_model }}_beg
+		{{ dbt_utils.group_by(2) }}
+
+		union 
+
+		select 
+			pernr,
+			date_change 
+		from {{ employee_model }}_end
+		{{ dbt_utils.group_by(2) }}
+
+			{% if not loop.last %}
+    		union
+			{% endif %} 
+
+		{% endif %} 
+	{% endfor %}
+	order by pernr, date_change
+),
+
+employee_date_ranges as (
+
+	select 
+		pernr,
+		date_change as begda,
+		lead(date_change, 1) over (partition by pernr order by date_change) as endda      
+	from unioned_dates			
+),
+
+employee_original_date_ranges as (
+
+	select
+		pernr,
+		begda,
+		case when endda = '99991231' then endda
+			else
+				--convert date strings to dates to avoid null records when joining on date ranges in the end model
+				{% if target.type in ('redshift', 'snowflake', 'snowflake-sap') %}
+				to_char(dateadd(day, -1, to_date(endda, 'YYYYMMDD')), 'YYYYMMDD') end as endda
+				{% elif target.type == 'postgres' %}
+				to_char(to_date(endda, 'YYYYMMDD') - interval '1 day', 'YYYYMMDD') end as endda
+				{% elif target.type == 'bigquery' %}
+				format_date('%Y%m%d', date_sub(parse_date('%Y%m%d', endda), interval 1 day)) end as endda
+				{% elif target.type == 'databricks' %}
+				date_format(date_sub(to_date(endda, 'yyyyMMdd'), 1), 'yyyyMMdd') end as endda
+				{% endif %}
```

Comment:
This macro is slightly different. Instead of adding the dates to remove the null date ranges, it's subtracting the date back so that we get the original end dates as they would appear in SAP. But I can create two separate macros here.

Reply:
This macro is slightly different from the one above (it's subtracting the date by one rather than adding one to return it to its original end date value), but I can create a separate one here.

Reply:
Changes applied

Reply:
Same question about possibly using dispatches. The dispatch logic allows the determining of which version of the macro to be used is built into dbt-core. However, the conditional logic is not as robust and is a bit trickier to maintain in the long run. 

If there were issues using the dispatch that were unforeseen then I am okay keeping the conditionals, but wanted to inquire if it would be possible to leverage the dispatch approach.

Reply:
@fivetran-joemarkiewicz Rewrote it now and I think it's working!